argoverse/evaluation/eval_detection.py

```diff
@@ -0,0 +1,271 @@
+                scale_error = dist_fn(dt_df, gt_df, DistFnType.SCALE)
+                orient_error = dist_fn(dt_df, gt_df, DistFnType.ORIENTATION)
+
+                error_types[dt_tp_indices, n_threshs : n_threshs + 3] = np.vstack(
+                    (trans_error, scale_error, orient_error)
+                ).T
+
+        return np.hstack((error_types, scores))
+
+    def summarize(self, data: defaultdict, cls_inst_map: defaultdict) -> defaultdict:
+        """Calculate and print the detection metrics.
+
+        Args:
+            data: The aggregated data used for summarization.
+            cls_inst_map: Map of classes to number of instances.
+
+        Returns:
+            summary: The summary statistics.
+        """
+        summary = defaultdict(list)
+        rec_interp = np.linspace(0, 1, self.dt_cfg.n_rec_samples)
+        num_ths = len(self.dt_cfg.sim_ths)
+        if not self.figures_fpath.is_dir():
+            self.figures_fpath.mkdir(parents=True, exist_ok=True)
+        for cls_name, cls_stats in data.items():
+            num_inst = cls_inst_map[cls_name]
+            ranks = cls_stats[:, -1].argsort()[::-1]
+            cls_stats = cls_stats[ranks]
+            for i, _ in enumerate(self.dt_cfg.sim_ths):
+                tp = cls_stats[:, i].astype(bool)
+
+                cumulative_tp = np.cumsum(tp, dtype=np.int)
+                cumulative_fp = np.cumsum(~tp, dtype=np.int)
+                cumulative_fn = num_inst - cumulative_tp
+
+                precs = cumulative_tp / (cumulative_tp + cumulative_fp + np.finfo(float).eps)
+                recs = cumulative_tp / (cumulative_tp + cumulative_fn)
+
+                precs = interp(precs)
+                prec_interp = np.interp(rec_interp, recs, precs, right=0)
+                ap_th = prec_interp.mean()
+                summary[cls_name] += [ap_th]
+
+            # AP Metric
+            ap = np.array(summary[cls_name][:num_ths]).mean()
+
+            # TP Error Metrics
+            tp_metrics = np.mean(cls_stats[:, num_ths : num_ths + 3], axis=0)
+            tp_metrics[2] /= np.pi / 2  # normalize orientation
+
+            # clip so that we don't get negative values in (1 - ATE)
+            cds_summands = np.hstack((ap, np.clip(1 - tp_metrics, a_min=0, a_max=None)))
+
+            # Ranking metric
+            cds = np.average(cds_summands, weights=self.dt_cfg.cds_weights)
+
+            summary[cls_name] = [ap, *tp_metrics, cds]
+            self.plot(rec_interp, prec_interp, cls_name)
+        logger.info("summary = %s" % summary)
+        return summary
+
+    def plot(self, rec_interp: np.ndarray, prec_interp: np.ndarray, cls_name: str) -> None:
+        """Plot and save the precision recall curve.
+
+        Args:
+            rec_interp: The interpolated recall data.
+            prec_interp: The interpolated precision data.
+            cls_name: Class name.
+        """
+        plt.plot(rec_interp, prec_interp)
```

Comment:
I think we might have to do something about the plot. Got an error `_tkinter.TclError: no display name and no $DISPLAY environment variable` (on a headless server)

Reply:
There seem to be a fix of like [2 line of code ](https://stackoverflow.com/questions/2801882/generating-a-png-with-matplotlib-when-display-is-undefined) but I don't know how that would effect others system.

Maybe we just add an option to disable ploting or something?

Reply:
@janesjanes, would you try adding `matplotlib.use("Agg")` to the top of the file? We could additionally add a flag, but this should allow for headless writing, regardless.

Reply:
The server is pulling directly from your repo currently, and i dont really have access to the file 😅  id imagine some people will use it on headless as well, right? maybe we can add a flag for headless so we can still save the figure?